runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,224 @@
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+
+import org.corfudb.protocols.logprotocol.MultiObjectSMREntry;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.StreamSubscriptionException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.ObjectsView;
+import org.corfudb.runtime.view.StreamOptions;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.runtime.view.stream.IStreamView;
+
+/**
+ * A simple thread based subscription engine where each subscriber or listener gets
+ * a thread which will listen on the tables of interest to it and moves at a rate
+ * which is the same as the consumption.
+ *
+ * Created by hisundar on 04/28/2020.
+ */
+@Slf4j
+public class StreamManager {
+
+    private static final int MAX_SUBSCRIBERS = 12; // Arbitrary sanity check value.
+
+    /**
+     * Corfu Runtime.
+     */
+    @Getter
+    private final CorfuRuntime runtime;
+
+    /**
+     * Map of StreamingSubscriptionThreads.
+     * <p>
+     * This is the map of all streaming subscriptions.
+     */
+    @Getter
+    private final Map<StreamListener, SubscriptionThread> subscriptions = new HashMap<>();
+
+    public StreamManager(@Nonnull CorfuRuntime runtime) {
+        this.runtime = runtime;
+    }
+
+    /**
+     * Subscribe to updates.
+     *
+     * @param streamListener   Client listener.
+     * @param namespace        Namespace of interest.
+     * @param tablesOfInterest Only updates from these tables will be returned.
+     * @param startAddress     Address to start the notifications from.
+     */
+    public synchronized <K extends Message, V extends Message, M extends Message>
+    void subscribe(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                   @Nonnull List<TableSchema<K, V, M>> tablesOfInterest, long startAddress) {
+        if (subscriptions.size() >= MAX_SUBSCRIBERS) {
+            throw new StreamSubscriptionException(
+                    "subscribe: Max " + MAX_SUBSCRIBERS + " hit! Consider consolidating "
+                            + streamListener);
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new StreamSubscriptionException(
+                    "StreamManager:subscriber already registered "
+                            + streamListener + ". Maybe call unregister first?");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}",
+                streamListener, startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
```

Comment:
These are not deamon threads, they can hang the JVM on exit.

Reply:
making them daemon threads create other problems, it will ends up masking issues where the vertical subscription thread which should be shutdown before JVM shutdown aren't told to exit. we saw this problem with a few consumers.

Reply:
The streamManager should have some shutdown/close logic where the threads are stopped properly and an exit callback is fired for the subscribers. Not exposing a close method and not managing the life-cycle inside the runtime is kind of problematic. Its not clear whose responsible for releasing those resources.

Reply:
just realized what you are describing as a problem is exactly the solution we want - we DO want to hang the JVM on exit otherwise corfu becomes responsible for an application level bug where the caller did not cleanly unsubscribe() and shutdown the subscription threads.
Callers are supposed to explicitly detect JVM shutdown and invoke unsubscribe() 
unsubscribe() is the shutdown/close logic.